File: ios/RNMBX/RNMBXImages.swift
import MapboxMaps

protocol RNMBXImageSetter : AnyObject {
  func addImage(name: String, image: UIImage, sdf: Bool?, stretchX: [[NSNumber]], stretchY: [[NSNumber]], content: [NSNumber]?, log: String) -> Bool
}

func hasImage(style: Style, name: String) -> Bool {
  #if RNMBX_11
  return style.imageExists(withId: name)
  #else
  return (style.styleManager.getStyleImage(forImageId: name) != nil)
  #endif
}

#if RNMBX_11
typealias StyleImageMissingPayload = StyleImageMissing
#endif

open class RNMBXImages : UIView, RNMBXMapComponent {
  
  @objc public weak var bridge : RCTBridge! = nil
  
  weak var style: Style? = nil

  var imageManager: ImageManager? = nil

  @objc
  public var onImageMissing: RCTBubblingEventBlock? = nil
  
  @objc
  public var images : [String:Any] = [:] {
    didSet {
      updateImages(images: images, oldImages: oldValue)
    }
  }
    
  var loadedImages : Set<String> = []
  
  var imageViews: [RNMBXImage] = []

  @objc
  public var nativeImages: [Any] = [] {
    didSet {
      nativeImageInfos = nativeImages.compactMap { decodeImage($0) }
    }
  }

  typealias NativeImageInfo = (name:String, sdf: Bool, stretchX:[(from:Float, to:Float)], stretchY:[(from:Float, to:Float)], content: (left:Float,top:Float,right:Float,bottom:Float)? );
  var nativeImageInfos: [NativeImageInfo] = []
  
  @objc public func addImageView(_ image: RNMBXImage) {
    imageViews.append(image)
  }

  @objc public func removeImageView(_ image: RNMBXImage) {
    imageViews.removeAll { $0 == image }
    image.images = nil
  }

  @objc open override func insertReactSubview(_ subview: UIView!, at atIndex: Int) {
    if let image = subview as? RNMBXImage {
      addImageView(image)
    } else {
      Logger.log(level:.warn, message: "RNMBXImages children can only be RNMBXImage, got \(optional: subview)")
    }
    super.insertReactSubview(subview, at: atIndex)
  }
  
  @objc open override func removeReactSubview(_ subview: UIView!) {
    if let image = subview as? RNMBXImage {
      removeImageView(image)
    }
    super.removeReactSubview(subview)
  }
  
  // MARK: - RNMBXMapComponent

  public func waitForStyleLoad() -> Bool {
    return false
  }
  
  public func addToMap(_ map: RNMBXMapView, style: Style) {
    self.style = style
    imageManager = map.imageManager
    map.images.append(self)
    
    self.addNativeImages(style: style, nativeImages: nativeImageInfos)
    self.addImages(style: style, images: images, oldImages: [:])
    self.addImageViews(style: style, imageViews: imageViews)
  }
  
  public func removeFromMap(_ map: RNMBXMapView, reason: RemovalReason) -> Bool {
    self.style = nil
    imageManager = nil
    // v10todo
    return true
  }
  
  func sameImage(oldValue: Any?, newValue: Any?) -> Bool {
    if let oldS = oldValue as? String, let newS = newValue as? String {
      return oldS == newS
    } else {
      return false
    }
  }

  func updateImages(images: [String:Any], oldImages: [String:Any]) {
    if let style = self.style {
      addImages(style: style, images: images, oldImages: oldImages)
    }
  }

  func addImages(style: Style, images: [String: Any], oldImages: [String:Any]) {
    var missingImages : [String:Any] = [:]
    
    // Add image placeholder for images that are not yet available in the style. This way
    // we can load the images asynchronously and add the ShapeSource to the map without delay.
    // The same is required when this ShapeSource is updated with new/added images and the
    // data references them. In which case addMissingImageToStyle will take care of loading
    // them in a similar way.
    //
    // See also: https://github.com/mapbox/mapbox-gl-native/pull/14253#issuecomment-478827792
    
    for name in images.keys {
      let newImage = oldImages[name] == nil
      if oldImages[name] == nil || loadedImages.contains(name) || !sameImage(oldValue: oldImages[name], newValue: images[name]) {
        let addPlaceholder = oldImages[name] == nil
        if !sameImage(oldValue: oldImages[name], newValue: images[name]) {
          missingImages[name] = images[name]
        } else {
          if !hasImage(style: style, name: name) {
            logged("RNMBXImages.addImagePlaceholder") {
              try style.addImage(placeholderImage, id: name, stretchX: [], stretchY: [])
              missingImages[name] = images[name]
            }
          }
        }
      }
    }
    
    if missingImages.count > 0 {
      RNMBXUtils.fetchImages(bridge, style: style, objects: missingImages, forceUpdate: true) { name, image in
        self.loadedImages.insert(name)
        self.imageManager?.resolve(name: name, image: image)
      }
    }
  }
  
  private func addImageViews(style: Style, imageViews: [RNMBXImage]) {
    imageViews.forEach { imageView in
      imageView.images = self
    }
  }
  
  internal func addMissingImageToStyle(style: Style, imageName: String) -> Bool {
    if let nativeImage = nativeImageInfos.first(where: { $0.name == imageName }) {
      addNativeImages(style: style, nativeImages: [nativeImage])
      return true
    }
    
    if let image = images[imageName] {
      addImages(style: style, images: [imageName: image], oldImages: [:])
      return true
    }
    return false
  }
  
  internal func sendImageMissingEvent(imageName: String, payload: StyleImageMissingPayload) {
    let payload = ["imageKey":imageName]
    let event = RNMBXEvent(type: .imageMissing, payload: payload)
    if let onImageMissing = onImageMissing {
      onImageMissing(event.toJSON())
    }
  }
  
  static func convert(stretch: [[NSNumber]], scale: CGFloat = 1.0) -> [(from: Float, to: Float)] {
    return stretch.map{ pair in
      return (from: pair[0].floatValue * Float(scale), to: pair[1].floatValue * Float(scale))
    }
  }
  
  static func convert(stretch: [(from: Float, to: Float)], scale: CGFloat = 1.0) -> [ImageStretches] {
    return stretch.map { v in ImageStretches(first: v.from * Float(scale), second: v.to * Float(scale)) }
  }
  
  static func convert(stretch: [[NSNumber]], scale: CGFloat = 1.0) -> [ImageStretches] {
    return convert(stretch: convert(stretch: stretch, scale: scale))
  }
  
  static func convert(stretch: [[NSNumber]], scale: Float = 1.0) -> [ImageStretches] {
    return convert(stretch: stretch, scale: CGFloat(scale))
  }


  static func convert(content: (left:Float, top:Float, right:Float, bottom:Float)?, scale: CGFloat) -> ImageContent? {
    guard let content = content else {
      return nil
    }
    
    return ImageContent(left:content.left*Float(scale), top:content.top*Float(scale), right:content.right*Float(scale), bottom:content.bottom*Float(scale))
  }

  static func convert(content: [NSNumber]?, scale: CGFloat = 1.0) -> (left:Float,top:Float,right:Float,bottom:Float)? {
    guard let content = content else {
      return nil
    }
    guard content.count == 4 else {
      Logger.log(level: .error, message: "Image content should have 4 elements got \(content)")
      return nil
    }
    return (
      left: content[0].floatValue*Float(scale),
      top: content[1].floatValue*Float(scale),
      right: content[2].floatValue*Float(scale),
      bottom: content[3].floatValue*Float(scale)
    )
  }
  
  static func convert(content: [NSNumber]?, scale: CGFloat = 1.0) -> ImageContent? {
    return convert(content: convert(content: content, scale: scale), scale: 1.0)
  }
  
  func decodeImage(_ imageNameOrInfo: Any) -> NativeImageInfo? {
    if let imageName = imageNameOrInfo as? String {
      return (name: imageName, sdf: false, stretchX:[],stretchY:[],content:nil)
    } else if let imageInfo = imageNameOrInfo as? [String:Any] {
      guard let name = imageInfo["name"] as? String else {
        Logger.log(level: .warn, message: "NativeImage: \(imageInfo) has no name key")
        return nil
      }
      var sdf = false
      var stretchX : [(from:Float, to:Float)] = []
      var stretchY : [(from:Float, to:Float)] = []
      if let sdfV = imageInfo["sdf"] as? NSNumber {
        sdf = sdfV.boolValue
      }
      
      if let stretchXV = imageInfo["stretchX"] as? [[NSNumber]] {
        stretchX = RNMBXImages.convert(stretch: stretchXV)
      }
      
      if let stretchYV = imageInfo["stretchY"] as? [[NSNumber]] {
        stretchY = RNMBXImages.convert(stretch: stretchYV)
      }
      
      var content : (left:Float, top:Float, right:Float, bottom:Float)? = nil
      if let contentV = imageInfo["content"] as? [NSNumber] {
        content = RNMBXImages.convert(content: contentV)
      }
      
      return (name: name, sdf: sdf, stretchX: stretchX, stretchY: stretchY, content: content)
    } else {
      Logger.log(level: .warn, message: "RNMBXImage.nativeImage, unexpected image: \(imageNameOrInfo)")
      return nil
    }
  }
  
  func addNativeImages(style: Style, nativeImages: [NativeImageInfo]) {
    for imageInfo in nativeImages {
      let imageName = imageInfo.name
      if  !hasImage(style: style, name: imageName) {
        if let image = UIImage(named: imageName) {
          logged("RNMBXImage.addNativeImage: \(imageName)") {
            try style.addImage(image, id: imageName, sdf: imageInfo.sdf,
                               stretchX: RNMBXImages.convert(stretch: imageInfo.stretchX, scale: image.scale),
                               stretchY: RNMBXImages.convert(stretch: imageInfo.stretchY, scale: image.scale),
                               content: RNMBXImages.convert(content: imageInfo.content, scale: image.scale)
            )
          }
          imageManager?.resolve(name: imageName, image: image)
        } else {
          Logger.log(level:.error, message: "Cannot find nativeImage named \(imageName)")
        }
      }
    }
  }
  
  lazy var placeholderImage : UIImage = {
    UIGraphicsBeginImageContextWithOptions(CGSize(width: 1, height: 1), false, 0.0)
    let result = UIGraphicsGetImageFromCurrentImageContext()!
    UIGraphicsEndImageContext()
    return result
  }()
}

extension RNMBXImages : RNMBXImageSetter {
  func addImage(name: String, image: UIImage, sdf: Bool?, stretchX: [[NSNumber]], stretchY: [[NSNumber]], content: [NSNumber]?, log: String) -> Bool
  {
     return logged("\(log).addImage") {
       if let style = style {
         try style.addImage(image,
                            id:name,
                            sdf: sdf ?? false,
                            stretchX: RNMBXImages.convert(stretch: stretchX, scale: image.scale),
                            stretchY: RNMBXImages.convert(stretch: stretchY, scale: image.scale),
                            content: RNMBXImages.convert(content: content, scale: image.scale)
         )
         return true
       } else {
         return false
       }
     } ?? false
 }
}

File: android/src/main/java/com/rnmapbox/rnmbx/components/images/RNMBXImages.kt
package com.rnmapbox.rnmbx.components.images

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.util.DisplayMetrics
import android.util.Log
import androidx.core.content.res.ResourcesCompat
import com.mapbox.bindgen.Expected
import com.mapbox.bindgen.None
import com.mapbox.maps.Image
import com.mapbox.maps.ImageContent
import com.mapbox.maps.ImageStretches
import com.mapbox.maps.MapboxMap
import com.mapbox.maps.Style
import com.rnmapbox.rnmbx.R
import com.rnmapbox.rnmbx.components.AbstractMapFeature
import com.rnmapbox.rnmbx.components.RemovalReason
import com.rnmapbox.rnmbx.components.mapview.RNMBXMapView
import com.rnmapbox.rnmbx.events.ImageMissingEvent
import com.rnmapbox.rnmbx.utils.BitmapUtils
import com.rnmapbox.rnmbx.utils.DownloadMapImageTask
import com.rnmapbox.rnmbx.utils.ImageEntry
import com.rnmapbox.rnmbx.v11compat.image.addStyleImage
import com.rnmapbox.rnmbx.v11compat.image.emptyImage
import com.rnmapbox.rnmbx.v11compat.image.toImageData
import com.rnmapbox.rnmbx.v11compat.image.toMapboxImage
import java.lang.Float.max
import java.nio.ByteBuffer
import java.util.AbstractMap
import kotlin.math.ceil

fun Style.addBitmapImage(imageId: String, bitmap: Bitmap, sdf: Boolean = false, stretchX: List<ImageStretches> = listOf(), stretchY: List<ImageStretches> = listOf(), content: ImageContent? = null, scale: Double = 1.0) : Expected<String, None> {
    val byteBuffer = ByteBuffer.allocate(bitmap.byteCount)
    bitmap.copyPixelsToBuffer(byteBuffer)
    return this.addStyleImage(
        imageId,
        (1.0/((160.0/bitmap.density)) * scale).toFloat() ,
        Image(bitmap.width, bitmap.height, byteBuffer.array().toImageData()),
        sdf,
        stretchX,
        stretchY,
        content,
    )
}

fun Style.addBitmapImage(nativeImage: NativeImage) : Expected<String, None> {
    val info = nativeImage.info
    return addBitmapImage(info.name, nativeImage.drawable.bitmap, info.sdf, info.stretchX, info.stretchY, info.content, info.getScaleOr(1.0))
}

data class ImageInfo(val name: String,  val scale: Double? = 1.0, val sdf: Boolean = false, val stretchX: List<ImageStretches> = listOf(),
                     val stretchY: List<ImageStretches> = listOf(), val content: ImageContent? = null, val width: Double? = null, val height: Double? = null)
{
    fun getScaleOr(default: Double): Double {
        return scale ?: default;
    }
}

data class NativeImage(val info: ImageInfo, val drawable: BitmapDrawable);

interface NativeImageUpdater {
    fun updateImage(imageId: String, bitmap: Bitmap, sdf: Boolean = false, stretchX: List<ImageStretches> = listOf(), stretchY: List<ImageStretches> = listOf(), content: ImageContent? = null, scale: Double = 1.0)
}

class RNMBXImages(context: Context, private val mManager: RNMBXImagesManager) : AbstractMapFeature(context), NativeImageUpdater {
    var mCurrentImages: MutableSet<String?>
    var mImageViews = mutableListOf<RNMBXImage>();
    private var mImages: MutableMap<String, ImageEntry>?
    private var mNativeImages = mutableMapOf<String, NativeImage>()
    private var mSendMissingImageEvents = false
    private var mMap: MapboxMap? = null

    fun setImages(images: List<Map.Entry<String, ImageEntry>>) {
        val newImages: MutableMap<String, ImageEntry> = HashMap()
        for ((key, value) in images) {
            val oldValue = mImages?.put(key, value)
            if (oldValue == null || value.uri != oldValue.uri) {
                newImages[key] = value
            }
        }
        if (mMap != null && mMap?.getStyle() != null) {
            addImagesToStyle(newImages, mMap!!)
        }
    }

    fun setNativeImages(nativeImages: List<NativeImage>) {
        val newImages: MutableMap<String, NativeImage> = HashMap()
        for (nativeImage in nativeImages) {
            val name = nativeImage.info.name
            val oldValue = mNativeImages.put(name, nativeImage)
            if (oldValue == null) {
                newImages[name] = nativeImage
            }
        }
        mMap?.let {
            if (it.getStyle() != null) {
                addNativeImagesToStyle(newImages, it)
            }
        }
    }

    fun setHasOnImageMissing(value: Boolean) {
        mSendMissingImageEvents = value
    }

    override fun removeFromMap(mapView: RNMBXMapView, reason: RemovalReason): Boolean {
        removeImages(mapView)
        mMap = null
        if (reason == RemovalReason.ON_DESTROY) {
            mNativeImages = mutableMapOf()
            mImages = HashMap()
            mCurrentImages = HashSet()
        }
        return super.removeFromMap(mapView, reason)
    }

    private fun removeImages(mapView: RNMBXMapView) {
        mapView.getStyle(object : Style.OnStyleLoaded {
            override fun onStyleLoaded(style: Style) {
                if (hasImages()) {
                    for ((key) in mImages!!) {
                        style.removeStyleImage(key)
                    }
                }
                if (hasNativeImages()) {
                    for ((key) in mNativeImages!!) {
                        style.removeStyleImage(key!!)
                    }
                }
            }
        })
    }

    private fun hasImages(): Boolean {
        return mImages != null && mImages!!.size > 0
    }

    private fun hasNativeImages(): Boolean {
        return mNativeImages != null && mNativeImages!!.size > 0
    }

    fun addMissingImageToStyle(id: String, map: MapboxMap): Boolean {
        val nativeImage = mNativeImages.get(id)
        if (nativeImage != null) {
            addNativeImages(listOf(nativeImage), map)
            return true
        }
        if (mImages != null) {
            val entry = mImages!![id]
            if (entry != null) {
                addRemoteImages(entry(id, entry), map)
                return true
            }
        }
        return false
    }

    fun addImagesToStyle(images: Map<String, ImageEntry>?, map: MapboxMap) {
        if (images != null) {
            addRemoteImages(ArrayList(images.entries), map)
        }
    }

    fun addNativeImagesToStyle(images: Map<String, NativeImage>, map: MapboxMap) {
        addNativeImages(images.values.toList(), map)
    }

    fun sendImageMissingEvent(id: String, map: MapboxMap) {
        if (mSendMissingImageEvents) {
            mManager.handleEvent(ImageMissingEvent.makeImageMissingEvent(this, id))
        }
    }

    private fun hasImage(imageId: String?, map: MapboxMap): Boolean {
        val style = map.getStyle()
        return style != null && imageId?.let { style.getStyleImage(it) } != null
    }

    override fun addToMap(mapView: RNMBXMapView) {
        super.addToMap(mapView)

        mImageViews.forEach {
            it.addToMap(mapView)
        }
        // Wait for style before adding the source to the map
        // only then we can pre-load required images / placeholders into the style
        // before we add the ShapeSource to the map
        mapView.getStyle(object : Style.OnStyleLoaded {
            override fun onStyleLoaded(style: Style) {
                val map = mapView.getMapboxMap()
                mMap = map
                addNativeImagesToStyle(mNativeImages, map)
                addImagesToStyle(mImages, map)
                // super.addToMap(mapView);

                mImageViews.forEach {
                    it.refresh()
                }
            }
        })
    }

    private fun addNativeImages(imageEntries: List<NativeImage>, map: MapboxMap) {
        val style = map.getStyle()
        if (style == null) return
        for (nativeImage in imageEntries) {
            val name = nativeImage.info.name
            if (!hasImage(name, map)) {
                val bitmap = nativeImage.drawable
                mMapView!!.imageManager.resolve(name, nativeImage.drawable)
                style.addBitmapImage(nativeImage)
                mCurrentImages.add(name)
            }
        }
    }

    private fun placeholderImageFor(info: ImageInfo): Image? {
        if (info.width != null && info.height != null) {
            val width = (info.width * info.getScaleOr(1.0)).toInt()
            val height = (info.width * info.getScaleOr(1.0)).toInt()

            return emptyImage(width, height)
        } else if (info.stretchX.isNotEmpty() || info.stretchY.isNotEmpty() || info.content != null) {
            var maxY = info.stretchY.maxOfOrNull { max(it.first, it.second) } ?: 0.0f
            var maxX = info.stretchX.maxOfOrNull { max(it.first, it.second) } ?: 0.0f
            if (info.content != null) {
                maxX = max(max(info.content.left, info.content.right), maxX)
                maxY = max(max(info.content.top, info.content.bottom), maxY)
            }
            return emptyImage(ceil(maxX).toInt()+1, ceil(maxY).toInt()+1)
        } else {
           return mImagePlaceholder?.toMapboxImage() ?: emptyImage(16, 16)
        }
    }

    private fun addRemoteImages(imageEntries: List<Map.Entry<String, ImageEntry>>?, map: MapboxMap) {
        val style = map.getStyle()
        if (style == null || imageEntries == null) return
        val missingImages: MutableList<Map.Entry<String, ImageEntry>> = ArrayList()

        val metrics = context.resources.displayMetrics
        val screenDensity = (metrics?.density ?: DisplayMetrics.DENSITY_DEFAULT).toDouble()

        // Add image placeholder for images that are not yet available in the style. This way
        // we can load the images asynchronously and add the ShapeSource to the map without delay.
        // The same is required when this ShapeSource is updated with new/added images and the
        // data references them. In which case addMissingImageToStyle will take care of loading
        // them in a similar way.
        //
        // See also: https://github.com/mapbox/mapbox-gl-native/pull/14253#issuecomment-478827792
        for (imageEntry in imageEntries) {
            if (!hasImage(imageEntry.key, map)) {
                val info = imageEntry.value.info;
                val placeholderImage = placeholderImageFor(info)
                placeholderImage?.let {
                    style.addStyleImage(imageEntry.key, it, info)
                }
                mCurrentImages.add(imageEntry.key)
            }
            // make image download even in case the URL changed
            missingImages.add(imageEntry)
        }
        if (missingImages.size > 0) {
            val task = DownloadMapImageTask(context, map, mMapView!!.imageManager)
            val params = missingImages.toTypedArray()
            task.execute(params)
        }
    }

    companion object {
        private var mImagePlaceholder: Bitmap? = null
        fun <K, V> entry(k: K, v: V): List<Map.Entry<K, V>> {
            return listOf(AbstractMap.SimpleEntry(k, v) as Map.Entry<K, V>)
        }
    }

    init {
        mCurrentImages = HashSet()
        mImages = HashMap()
        mNativeImages = HashMap()
        if (mImagePlaceholder == null) {
            mImagePlaceholder = BitmapUtils.getBitmapFromDrawable(ResourcesCompat.getDrawable(context.resources, R.drawable.empty_drawable, null))
        }
    }

    override fun updateImage(imageId: String, bitmap: Bitmap, sdf: Boolean, stretchX: List<ImageStretches>, stretchY: List<ImageStretches>, content: ImageContent?, scale: Double)
    {
        mMap?.getStyle()?.let {
            it.addBitmapImage(imageId, bitmap, sdf, stretchX, stretchY, content, scale);
        }
    }
}